lm/vocabulary.py

```diff
@@ -1,6 +1,125 @@
+
+    If <synt> tokens are provided in a lemma, they will be used instead of <orth>
+
+    CAUTION:
+    Be aware of: https://github.com/rwth-i6/returnn/issues/1245 when using Returnn's LmDataset
+    """
+
+    def __init__(
+        self, bliss_lexicon: tk.Path, count_ordering_text: Optional[tk.Path] = None
+    ):
+        self.bliss_lexicon = bliss_lexicon
+        self.count_ordering_text = count_ordering_text
+
+        self.out_vocab = self.output_path("lm.vocab.txt")
+        self.out_vocab_size = self.output_var("vocab_size")
+        self.out_unknown_token = self.output_var("unknown_token")
+
+        self.out_vocabulary_object = LmIndexVocabulary(
+            vocab=self.out_vocab,
+            vocab_size=self.out_vocab_size,
+            unknown_token=self.out_unknown_token,
+        )
+
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+    def run(self):
+        lex = Lexicon()
+        lex.load(self.bliss_lexicon.get_path())
+        wordlist = []
+        sentence_begin = None
+        sentence_end = None
+        sentence_boundary = None
+        unknown = None
+
+        for lemma in lex.lemmata:
+            if lemma.special == "sentence-begin":
+                sentence_begin_list = (
+                    lemma.synt if lemma.synt is not None else lemma.orth
+                )
+                assert len(sentence_begin_list) == 1
+                sentence_begin = sentence_begin_list[0]
+            elif lemma.special == "sentence-end":
+                sentence_end_list = lemma.synt if lemma.synt is not None else lemma.orth
+                assert len(sentence_end_list) == 1
+                sentence_end = sentence_end_list[0]
+            elif lemma.special == "sentence-boundary":
+                sentence_boundary_list = (
+                    lemma.synt if lemma.synt is not None else lemma.orth
+                )
+                assert len(sentence_boundary_list) == 1
+                sentence_boundary = sentence_boundary_list[0]
+            elif lemma.special == "unknown":
+                unknown_list = lemma.synt if lemma.synt is not None else lemma.orth
+                assert len(unknown_list) == 1
+                unknown = unknown_list[0]
+            elif lemma.synt is not None:
+                wordlist.extend(lemma.synt)
+            else:
+                wordlist.extend(lemma.orth)
+
+        assert sentence_boundary is not None or (
+            sentence_begin is not None and sentence_end is not None
+        )
+        assert unknown is not None
+
+        if self.count_ordering_text is not None:
+            word_set = set(wordlist)
+            word_counts = {}
+            for w in wordlist:
+                word_counts[w] = 0
+            with uopen(self.count_ordering_text, "rt") as f:
+                for line in f.readlines():
+                    for word in line.strip().split(" "):
+                        if word in word_set:
+                            word_counts[word] += 1
```

Comment:
`word_set` is not needed here, as the keys of a dict already are a set (with the same inplications for runtime cost)

```suggestion
            word_counts = {}
            for w in wordlist:
                word_counts[w] = 0
            with uopen(self.count_ordering_text, "rt") as f:
                for line in f.readlines():
                    for word in line.strip().split(" "):
                        if word in word_counts:
                            word_counts[word] += 1
```

Reply:
but this comment got me thinking: should the `wordlist` not rather be a `wordset`? I get it, that you would like to impose some kind of ordering and in the absence of a text to get word counts from you would rely on the order of appearance in the original lexicon. 
But what if there are homonyms (multiple lemmata with same orth or synth) in the lexicon? Then the same word would appear twice in the list.

Maybe the best solution would be to have a dict (OrderedDict) to begin with and during the `for lemma in lex.lemmata` loop the counts are initialized as 0. Then if `self.count_ordering_text is not None` the counts are updated and the `wordlist` is created as you have below and else ```wordlist = word_counts.keys()``` relying on the OrderedDict to keep the order of appearance.

Reply:
But this will remove duplicates or not?
```python
            for w in wordlist:
                word_counts[w] = 0
                ```

Reply:
yes, but only in the case that a `self.count_ordering_text` is given, otherwise this part of the code is not used and there could be duplicates.

Reply:
maybe like this?